### PR TITLE
Reland "[Windows] Move to FlutterCompositor for rendering"

### PR DESCRIPTION
**Pull request by loic-sharma**
## Reland

https://github.com/flutter/engine/pull/48849 was reverted as it incorrectly expected to receive always 1 layer. However, the engine will present 0 layers on an ["empty" app](https://github.com/flutter/flutter/blob/6981fe6fd3aacb95bfc4a6c427ee5d493f43c5dc/dev/integration_tests/ui/lib/empty.dart#L8-L19). This pull request is split into two commits:

1. https://github.com/flutter/engine/pull/49262/commits/df604a16c0047e26ae30df6a9c83b8fb7e5572fb is the original pull request, unchanged
2. https://github.com/flutter/engine/pull/49262/commits/c30b36928981e454a9caf692767e709757227bb2 adds the ability to "clear" the view if the engine presents 0 layers

## Original pull request description

This migrates the Windows embedder to `FlutterCompositor` so that the engine renders off-screen to a framebuffer instead of directly onto the window's surface. This will allow us to support platform views and multiple views on Windows.

Addresses https://github.com/flutter/flutter/issues/128904

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/compositor_opengl.cc [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    gl_.reset();
+    return false;
+  }
+
+  format_ = GetSupportedTextureFormat(gl_->GetDescription());
+  is_initialized_ = true;
+  return true;
+}
+
+bool CompositorOpenGL::ClearSurface() {
```
Comment (loic-sharma) [inline shell/platform/windows/compositor_opengl.cc:172]: This clears previously presented layers if the engine presents with 0 layers

File: shell/platform/windows/compositor_software.cc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+bool CompositorSoftware::Present(const FlutterLayer** layers,
+                                 size_t layers_count) {
+  if (!engine_->view()) {
+    return false;
+  }
+
+  // Clear the view if there are no layers to present.
+  if (layers_count == 0) {
+    return engine_->view()->ClearSoftwareBitmap();
```
Comment (loic-sharma) [inline shell/platform/windows/compositor_software.cc:48]: This clears previously presented layers if the engine presents with 0 layers

**Comment by dkwingsmt** [comment]
This looks good to me, but I'd like people more specialized in these code to approve it.

File: shell/platform/windows/compositor.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+//
+// 1. Create backing stores used for rendering Flutter content
+// 2. Composite and present Flutter content and platform views onto a view
+//
+// Platform views are not yet supported.
+class Compositor {
+ public:
+  virtual ~Compositor() = default;
+
+  // Create a backing store used for rendering Flutter content.
```
Comment (cbracken) [inline shell/platform/windows/compositor.h:24]: nit: 
```suggestion
  // Creates a backing store used for rendering Flutter content.
```

**Review by cbracken — APPROVED** [review]
Thanks for breaking this up into two pieces. Looks great! Missed the doc nitpick on the first one but other than that this looks fantastic; I've got no useful review comments other than lgtm!

LGTM stamp from a Japanese personal seal

**Review by yaakovschectman — APPROVED** [review]
LGTM, looking forward to it landing

File: shell/platform/windows/compositor_opengl.cc [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  // Clear the view if there are no layers to present.
+  if (layers_count == 0) {
+    // Normally the compositor is initialized when the first backing store is
+    // created. However, on an empty frame no backing stores are created and
+    // the present needs to initialize the compositor.
+    if (!is_initialized_ && !Initialize()) {
+      return false;
+    }
+
+    return ClearSurface();
```
Comment (chinmaygarde) [inline shell/platform/windows/compositor_opengl.cc:109]: Clearing the surface requires a context to be made current. In this code path, you only do it after the call to ClearSurface. Is the context is not already previously current, this call will fail.

Comment (chinmaygarde) [reply shell/platform/windows/compositor_opengl.cc:109]: Oh, nevermind. You explicitly make another call in ClearSurface.

File: shell/platform/windows/compositor_opengl.cc [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+  // Resize the surface if needed.
+  engine_->view()->OnEmptyFrameGenerated();
+
+  if (!engine_->surface_manager()->MakeCurrent()) {
+    return false;
+  }
+
+  gl_->ClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+  gl_->Clear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
```
Comment (chinmaygarde) [inline shell/platform/windows/compositor_opengl.cc:183]: I'd clear all buffers. So include `GL_STENCIL_BUFFER_BIT` too. It is a no-op if the framebuffer doesn't contain one of these attachments.

File: shell/platform/windows/compositor_opengl.cc [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+  if (!engine_->surface_manager()->MakeCurrent()) {
+    return false;
+  }
+
+  auto source_id = layers[0]->backing_store->open_gl.framebuffer.name;
+  gl_->BindFramebuffer(GL_READ_FRAMEBUFFER, source_id);
+  gl_->BindFramebuffer(GL_DRAW_FRAMEBUFFER, kWindowFrameBufferId);
+
+  gl_->BlitFramebuffer(0,                    // srcX0
```
Comment (chinmaygarde) [inline shell/platform/windows/compositor_opengl.cc:139]: What version of GL are you using as the base? BlitFramebuffer won't be available on ES2 and you'll have to resort to render to texture instead.

Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc:139]: From my understanding, [ANGLE supports GLES 3.0 if Direct 3D 11 is available](https://github.com/google/angle#level-of-opengl-es-support-via-backing-renderers), which it is on Windows 7 and higher.

File: shell/platform/windows/compositor_opengl.cc (hunk 5 above)
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc]: Fixed!

File: shell/platform/windows/compositor_opengl.cc (hunk 4 above)
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc:109]: Yup that's correct. Marking as resolved

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**